### ISSUE TITLE
Lighthouse in DevTools errors/hangs on installability checks

### ISSUE THREAD

**Issue by connorjclark** [GoogleChrome/lighthouse#13396]
Latest:

We have a fix that should land in Canary ~soon: chromium-review.googlesource.com/c/chromium/src/+/3318838

We'll request a merge to Chrome 97 (scheduled to release Jan 4).

In the meantime, you can avoid the hanging by disabling Clear Storage
![image](https://user-images.githubusercontent.com/4071474/145088493-5b1acec7-6b82-4712-96f6-2d9fd91566c1.png)

--------

Before 98.0.4711.0, some sites with service workers would hang indefinitely while determining if there were any installability errors. With 98.0.4711.0, hanging should occur less often ([these issues were fixed](https://github.com/GoogleChrome/lighthouse/issues/13273#issuecomment-958576740)), a report will be generated and this error will be seen in the installable audit: `Lighthouse could not determine if there was a service worker. Please try with a newer version of Chrome.`. This message is aspirational, as there is not _yet_ a fix for the underlying cause.

However, there are still URLs that will hang indefinitely:

- https://www.enricofromrome.eu/edb6/
- https://restaurantguru.com/
- https://club.triboon.net/

[Debugging](https://github.com/GoogleChrome/lighthouse/issues/13273#issuecomment-960001555) shows that the target for the service worker is never resumed.

Some URLs that will show the new `could not determine` message (and not hang):

- https://wheresthebus.com/
- https://naughty-yonath-344d6c.netlify.app/
- https://periodex.co/
- https://m.warsztat24.pl/
- https://www.oklahomadriver.com/

related crbug: https://bugs.chromium.org/p/chromium/issues/detail?id=1276339

**Comment by chrisbennett-Bene** [comment]
Tried on Canary 98.0.4719.0. Still won't complete and didn't get a nice error message, but could at least cancel and see console, unlike in past. Error pops at the moment Lighthouse stalls.

`
Service worker registration failed or was blocked by your browser settings:  DOMException: Failed to register a ServiceWorker for scope ('https://www.alfresco-bar-bistro.com.au/') with script ('https://www.alfresco-bar-bistro.com.au/sw-6.2.0.v1.00.js'): Operation has been aborted
`

**Comment by mikkopori** [comment]
Tried on Version 98.0.4710.4 (Official Build) dev (64-bit) on Ubuntu

and also getting hanged with this in the console:

`Error during service worker registration: DOMException: Failed to register a ServiceWorker for scope ('http://localhost:3000/') with script ('http://localhost:3000/sw.js'): Operation has been aborted`

Was able to cancel also.

**Comment by theshanergy** [comment]
I am getting "

**Comment by theshanergy** [comment]
I am getting "FID issue: longer than 100ms (mobile)" errors site-wide on Google Search Console Core Web Vitals corresponding to this bug appearing. I believe it to be related to the service worker failing to load though I am not certain. Can anyone else confirm whether they are seeing this in their Search Console at the moment?

**Comment by connorjclark** [comment]
```sh
# python bisect-builds.py -o --verify-range -a mac -g 857950 -b 929940 --use-local-cache --not-interactive -c "bash $PWD/scripts/cdt-manifest-install-errors-hang/script.sh %p"

LH=~/src/lighthouse

export CHROME_PATH=$1

# brew install coreutils

cd $LH
pwd
if gtimeout 30 yarn run-devtools 'https://www.enricofromrome.eu/edb6/'; then
  echo "ok $1"
  exit 0
else
  echo "bad $1"
  exit 1
fi
```

result: https://chromium.googlesource.com/chromium/src/+log/1abcb14aea6f33b2f0e09fef09755d189b426f5f..38d1dcf838c6a49d59244c768d57638c7e34a7b8

Does not hang with this Chrome flag: `--disable-features=PlzServiceWorker`

**Comment by paulirish** [comment]
> Does not hang with this Chrome flag: `--disable-features=PlzServiceWorker`

HUGE FIND. 

This is definitely key. We've since followed up with some chromium folks to better understand what its all about.

**Comment by paulirish** [comment]
I've added some [basic printf logging to the installability manager](https://gist.github.com/paulirish/45938e75e395ace53c6006f86ff8599c) in Chromium to better understand what's happening.

The repro is basically: open https://caltrainschedule.io/ . open devtools. lighthouse panel. just pwa. hit analyze.  this will hit our protocol_timeout case.

But now with Chromium logging there's something more interesting going on.

### In a happy run (with PlzServiceWorker) it looks like this:

- lighthouse loads up about:blank twice. each time, chromium's `AppBannerManager::RequestAppBanner` is called which gathers [just the eligibility items](https://source.chromium.org/chromium/chromium/src/+/main:components/webapps/browser/banners/app_banner_manager.cc;l=62-66;drc=32f574f6df1d37b0d9ed765fe138270414d79e49). (AFAIK, this is done fairly eagerly w/ every load in case chrome wants to change the omnibox color to matches the manifest theme color).
- then LH has chrome navigate to the PWA (caltrainsched) and now chromium again automatically runs `AppBannerManager::RequestAppBanner` which is done in 100ms.
- because Chrome thinks this may potentially be installable, it continues with the rest of the installability checks aka `AppBannerManager::PerformInstallableWebAppCheck`. This does look at the serviceworker (the most involved check).
   - PerformInstallableWebAppCheck takes 200ms. Within that, the serviceworker inspection takes 25ms. 
- 7 seconds after all that, along comes Lighthouse very curious about the installability situation. it calls the `Page.getInstallabilityErrors` CDP method and that triggers a fresh call to `InstallableManager::GetData` that wants the exact information that `AppBannerManager` had just received (7s ago). 
- InstallableManager obliges and returns the cached data and 1ms later it's sent back over the protocol to Lighthouse who is so happy.  It's an empty array cuz this PWA is indeed installable. :)


### in a sad, hanging run though:

- same first 3 bullets.
   - HOWEVER.... the 3rd bullet (`PerformInstallableWebAppCheck`) hasn't finished yet when LH's CDP request comes along. It's stuck on `CheckServiceWorker()`. 
   - This time, `PerformInstallableWebAppCheck` takes 38 seconds. Within that, the serviceworker inspection takes 37 seconds.
- CDP eventually returns a response, but its been more than LH's `DEFAULT_PROTOCOL_TIMEOUT` of 30 seconds. So LH has no idea.

Also different this time: 

- After 37s of `CheckServiceWorker()` shenanigans it goes down a new `OnWaitingForServiceWorker()` route. That ends up going back into `CheckServiceWorker()` but this time it's done in ~10ms.
- `Uncaught (in promise) AbortError: Failed to register a ServiceWorker for scope ('https://caltrainschedule.io/') with script ('https://caltrainschedule.io/sw.js'): Operation has been aborted",` in the console.
   - because the SW is in a failure state, the CDP method includes that in the installabilityErrors payload: `['no-matching-service-worker]`

**Comment by connorjclark** [comment]
current theory is that the SW target is paused. Targets are resumed when we attach but for some reason we don't get the opportunity to resume the target dealing with this SW network fetch, which now happens in the browser. If you turn off the pausing on autoAttach in target managers on frame attached, then the hang does not happen ...

**Comment by connorjclark** [comment]
If I avoid our usage of `Storage.clearDataForOrigin` there is no hanging! So that seems important.

I found that when replaying a log of CDT protocol commands (from a hanging repro), and filtering out all the methods I could while keeping the hanging reproduction.

**Comment by connorjclark** [comment]
We have a fix that should land in Canary ~soon: https://chromium-review.googlesource.com/c/chromium/src/+/3318838

We'll request a merge to Chrome 97 (scheduled to release Jan 4).

**Comment by connorjclark** [comment]
Confirmed with latest Canary 98.0.4755.0 that all the URLs in the original post work, and none return the temporary error message.

Will keep this open to track the merge into Chrome 97 and (maybe) 96.

**Comment by paulirish** [comment]
Merged to 97: https://bugs.chromium.org/p/chromium/issues/detail?id=1276339#c30

I don't see any pending stuff on m96 merge tho.

**Comment by connorjclark** [comment]
Fixed reach stable.

**Comment by enricofromrome** [comment]
No changes in my software ('https://www.enricofromrome.eu/edb6/') and now, on chrome 97 it works. Thank you
